src/puppetlabs/puppetdb/utils.clj

```diff
@@ -470,3 +470,14 @@
   "remove all nil-valued keys from a map"
   [m]
   (into {} (filter val m)))
+
+(defn wait-for-ref-state [ref ms pred]
+  (let [watch-key wait-for-ref-state
+        finished? (promise)
+        handle-state #(when (pred %) (deliver finished? true))]
+    (add-watch ref watch-key (fn [_ _ _ new] (handle-state new)))
+    (try
+      (handle-state @ref)
```

Comment:
I'm wondering if there is anyway for this to return true even though there is a gc or other event we're wanting to wait for in progress before the timeout expires. In the gc case when we set the atom to :stopping right before calling wait-for-ref-state won't that cause our atom to have :stopping in the set so that the handle-state call in the try block will deliver true everytime without waiting for the timeout, or at least in most cases?

Reply:
On second thought, maybe this is already handled because if gc was going on you would get a set #{:stopping :collecting-garbage} and then the pred would return false and we'd still be waiting on an update from the watcher.

Reply:
Right -- all of this rests on the fact that the atom's state changes are...atomic, so that the state can only be #{}, #{:stopping}, or #{:stopping :collecting-garbage}, and also the fact that the :stopping state is "terminal", i.e. once added, it's never removed -- and if the garbage collection task sees :stopping, it bails out.

There is still a narrow period between when the gc task decides not to run (the if/when), and when the top-level enclosing function returns (presumably killing the lingering thread, given the bug), but we're ignoring that since it's not causing any noise, and it shouldn't be able to do any real harm, since all our state is in durable stores.